modules/flann/include/opencv2/flann/heap.h

```diff
@@ -49,117 +54,189 @@ namespace cvflann
+        heap.reserve(capacity);
+    }
 
     /**
-     * Insert a new element in the heap.
+     * \brief Insert a new element in the heap.
      *
      * We select the next empty leaf node, and then keep moving any larger
      * parents down until the right location is found to store this element.
      *
-     * Params:
-     *     value = the new element to be inserted in the heap
+     * \param value the new element to be inserted in the heap
      */
     void insert(T value)
     {
         /* If heap is full, then return without adding this element. */
-        if (count == length) {
+        if (size() == capacity()) {
             return;
         }
 
         heap.push_back(value);
-        static CompareT compareT;
-        std::push_heap(heap.begin(), heap.end(), compareT);
-        ++count;
+        static greater_equal comp;
+        std::push_heap(heap.begin(), heap.end(), comp);
     }
 
-
-
     /**
-     * Returns the node of minimum value from the heap (top of the heap).
+     * \brief Returns the node of minimum value from the heap (top of the heap).
      *
-     * Params:
-     *     value = out parameter used to return the min element
-     * Returns: false if heap empty
+     * \param[out] value parameter used to return the min element
+     * \returns false if heap empty
      */
     bool popMin(T& value)
     {
-        if (count == 0) {
+        if (empty()) {
             return false;
         }
 
         value = heap[0];
-        static CompareT compareT;
-        std::pop_heap(heap.begin(), heap.end(), compareT);
+        static greater_equal comp;
+        std::pop_heap(heap.begin(), heap.end(), comp);
         heap.pop_back();
-        --count;
 
         return true;  /* Return old last node. */
     }
+
+    /**
+     * \brief Returns a shared heap for the given memory pool ID.
+     *
+     * It constructs the heap if it does not already exists.
+     *
+     * \param poolId a user-chosen hashable ID for identifying the heap.
+     *     For thread-safe operations, using current thread ID is a good choice.
+     * \param capacity heap maximum capacity
+     * \param timeout remove heaps that were not reused for more than specified value (in seconds)
+     * \returns pointer to the heap
+     */
+    template <typename HashableT>
+    static std::shared_ptr<Heap<T>> getPooledInstance(
+        const HashableT& poolId, const int capacity, const int timeout = 30)
+    {
+        static std::mutex mutex;
+        const std::lock_guard<std::mutex> lock(mutex);
+
+        struct HeapMapValueType {
+            std::shared_ptr<Heap<T>> heapPtr;
+            std::chrono::steady_clock::time_point lastTimeUsed;
+        };
+        typedef std::unordered_map<HashableT, HeapMapValueType> HeapMapType;
+
+        static HeapMapType heapsPool;
+        typename HeapMapType::iterator heapIt = heapsPool.find(poolId);
+
+        const std::chrono::steady_clock::time_point& timeNow = std::chrono::steady_clock::now();
```

Comment:
OpenCV algorithms should be reproducible.
So time-based code should be avoided.
Please replace this with some "iteration counter".